Avoid downloading gcd if gcloud is installed and gcd version matches v1beta2

mziccard opened the pull request:
This PR checks whether `gcloud` and `gcd` are already installed, as per issue https://github.com/GoogleCloudPlatform/gcloud-java/issues/123. If `gcd` is installed and its version matches `v1beta2` then we use it, otherwise we download a matching version.

jgeewax commented:
/cc @aozarov :)

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -62,17 +64,85 @@
   public static final String DEFAULT_PROJECT_ID = "projectid1";
   public static final int PORT = 8080;
   private static final String GCD = "gcd-v1beta2-rev1-2.1.2b";
```
Lets make GCD be based on GCD_VERSION (GCD_VERSION + "...").
Feel free to find a better name for "GCD" and consider a constant for build number ("rev1-2.1.2.b").

jgeewax replied:
I'm actually -1 to that one -- we're already seeing some weird version numbers, which are really versions + revisions? Seems like just saying "this is the name of the GCD binary" (`gcd-v1beta2-revX-y.z.a`) is enough, right? I can't see ever using `'gcd'` separately from the `'v1-beta2-rev....'` string...

@mziccard - thoughts?

aozarov replied:
The only thing I was suggesting is to re-use constants as in this case 2 are having the `"v1beta2"` string.
These are internal constants, and I think It would make it more readable and maintainable.
Something like:
```java
private static final String GCD_VERSION = "v1beta2";
private static final String GCD_BUILD = "rev1-2.1.2b";
private static final String GCD = "gcd-" + GCD_VERSION + "-" + GCD_BUILD;
private static final String GCD_FILENAME = GCD + ".zip";

```

jgeewax replied:
Yea -- that was what I thought, and was wondering where you could safely re-use `GCD_VERSION` (or any of the others). 

Unless we have a good example, how about leaving as a string, and splitting apart when needed? It's a super simple refactor and means we have a good reason to do it...

aozarov replied:
It is not about safely re-use. it is about not forgetting to update "v1beta2" in one place but not the other.

jgeewax replied:
Agreed and understood. My question is: "Where else do we currently use the variable `GCD_VERSION` that means the same thing conceptually?"

For example, the `"v1beta2"` string we use for API calls doesn't really mean the same thing as the `GCD_VERSION` variable here, right? I wouldn't expect LocalGcdHelper's variables to be used in the Datastore implementation...

mziccard replied:
I think the point is that we use `GCD_VERSION` constant in LocalGcdHelper's code to check that the installed gcd version matches the one we require. If not, we download `GCD` which itself contains `GCD_VERSION` as substring. Doing as @aozarov proposes would require only changing `GCD_VERSION` and `GCD_BUILD` the day we need to update to a new gcd/API version. 

I think we have more chances of remembering to update `GCD_VERSION` and `GCD_BUILD` (as their names make sense) rather than `GCD_VERSION` and `GCD` where in `GCD` we need to change both version and build numbers.

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -62,17 +64,85 @@
+
+  }
+  
+  private static Path executablePath(String cmd) {
```
Does this work for Windows as well? (e.g I am not sure if gcloud is actually named gcloud.cmd on Windows).
An alternative could be do do in Java something like:
 gcloud info | grep "Installation Root" (and refactor the Process execute bellow to be common).

mziccard replied:
You are right. Gcloud on windows should be simply `gcloud`, not `gcloud.cmd`. I do not have a win machine to try ;(

aozarov replied:
I think you can use GCE with windows for testing.

jgeewax replied:
@mziccard : Ping me your work project ID and I can whitelist it for you :)

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -62,17 +64,85 @@
+  private static String installedGcdVersion() throws IOException, InterruptedException {
+    ProcessBuilder processBuilder = new ProcessBuilder()
+      .redirectErrorStream(true);
+    if (isWindows()) {
```
Lets refactor the Process invoking code to be shared between here, and gcd interaction (and maybe gcloud info) if possible.

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -62,17 +64,85 @@
+      try {
+        String installedVersion = installedGcdVersion();
+        if (installedVersion != null && installedVersion.startsWith(GCD_VERSION)) {
+          return installedGcdPath;
```
It would have been nice if code was already using debug/verbose level logs for "fetching", "using from cache" and even more so now "using sdk", "Found SDK but version but version mismatch",...

mziccard replied:
I can surely add some logging (and it would be useful to understand what's happening), the only problem I see is that logs will appear in the middle of test results.

aozarov replied:
Default log level is INFO, right? I was thinking of something finer that would not triggered by default for tests.

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -62,17 +64,85 @@
+  
+  private static String installedGcdVersion() throws IOException, InterruptedException {
+    ProcessBuilder processBuilder = new ProcessBuilder()
+      .redirectErrorStream(true);
```
this didn't fit in a 100 column line?

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -62,17 +64,85 @@
+    try (BufferedReader reader = 
+            new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+      for (String line = reader.readLine(); line != null; line = reader.readLine()) {
+        if (line.startsWith("gcd-emulator")) {
```
"gcd-emulator "?

mziccard replied:
When `gcloud version` is run the version of `gcd` appears as:
> gcd-emulator v1beta2-3.0.0

aozarov replied:
Yes the " " suffix is for protecting against future gcd-emulator-lib ... or keys with similar pattern.

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -157,21 +239,26 @@ public void start() throws IOException, InterruptedException {
     if (isWindows()) {
-      processBuilder.command("cmd", "/C", "gcd.cmd", "create", "-p", projectId, projectId);
-      processBuilder.redirectOutput(new File("NULL:"));
+      Path gcdAbsolutePath = executablePath.toAbsolutePath().resolve("gcd.cmd");
```
resolving executable could be done once and then call a common invoke process.

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -157,21 +239,26 @@ public void start() throws IOException, InterruptedException {
+      processBuilder.command("cmd", "/C", gcdAbsolutePath.toString(), "create",
+          "-p", projectId, projectId);
     } else {
-      processBuilder.redirectOutput(new File("/dev/null"));
```
Why did you get rid of the output redirect?

aozarov commented:
Thanks Marco, few small comments.

mziccard commented:
@aozarov First, thanks for the feedback. I refactored the process invocation code (added a class to wrap it) and added some fine-grained logging. I also fixed your other minor suggestions. 
It took me ages to test this on windows as I think I stumbled upon an issue with the `gcd.cmd` script (https://github.com/GoogleCloudPlatform/gcloud/issues/37).

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -157,38 +354,49 @@ public void start() throws IOException, InterruptedException {
+    if (log.isLoggable(Level.FINE)) {
+      log.log(Level.FINE, "Creating datastore for the project: {0}", projectId);
+    }
+    Process temp = gcdCreateCommand.command(gcdAbsolutePath.toString(), "create", "-p", projectId,
```
The preferred wrapping would be:
```java
    Process temp =
        gcdCreateCommand.command(gcdAbsolutePath.toString(), "create", "-p", projectId, projectId)
            .redirectErrorInherit()
            .directory(gcdPath)
            .redirectOutputToNull()
            .start();
```

see: http://google.github.io/styleguide/javaguide.html#s4.5.1-line-wrapping-where-to-break
and https://github.com/google/styleguide for formatters.

jgeewax replied:
can we also use a better name than "temp" ?

aozarov commented on gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/LocalGcdHelper.java:
```diff
@@ -115,6 +212,84 @@ public static ProcessStreamReader start(Process process, String blockUntil) thro
+    }
+    
+    public static CommandWrapper cmd() {
+      CommandWrapper wrapper = new CommandWrapper();
```
Would it be cleaner to have the constructor call isWindows() and initialize prefix and nullFilename?
bash() and cmd() need to make sure they initialize things correctly (e.g. nullFilename) to avoid NPE in builder() line 260.

I know that it means that in one case there are basically twice the isWindows() checks. once for picking the script name and one here but I think it would simplify the code of the callers. just use contractor or single factory method.

If you don't like that then maybe at least change redirectOutputToNull to redirectOutput(String or File) and builder() method redirect if not null.
check if string is

mziccard replied:
I was wondering which was the best solution when I implemented it. I went for the two factory methods as a `isWindows` check was already in the caller. But you are right, it's nicer to have a single factory method.

aozarov commented:
Looks good. 2 small changes and I think its ready to be merged.

aozarov commented:
Nice.  Thank you!
